Forward compatibility HIP (#370)
helm/community: hips/hip-XXXX.md

---
hip: 9999
title: "Forward compatibility: Chart.yaml minimumHelmVersion"
authors: [ "George Jenkins <gvjenkins@gmail.com>" ]
created: "2024-11-18"
type: "feature"
status: "draft"
---

## Abstract

> This HIP proposes supporting a `minimumHelmVersion` field in `Chart.yaml`.

Author: Happy to call the field `HelmVersion` (naming is hard :) )

---

I don't think we should/need to support anything for than a single, implied to be the minimum version, constraint however. I've addressed this here:

https://github.com/helm/community/pull/370/files#diff-e94b50b2c7a2d45c81db342448ae78950464d77488a62b315a2b715549fe1a95R95

By enabling this feature, an error can be raised if the Helm version used to operate on the chart is below the declared version.
This allows Helm to provide forward compatibility guarantees for Helm features/functionality over time.

## Motivation

Helm has no mechanism for a chart to declare the minimal version of Helm required for the chart to install/update correctly.
As such, it is invalid to release a chart that utilizes features/fixes included in newer versions of Helm.

At best, any incompatibility will be detected and there will be an explicit failure (and the user will be notified with an error).
But potentially an incompatibility may go undetected, and no hard error will be presented.

While the hard-failure case is better, it still requires the user to debug and realize the failure is due to a version mismatch.

> The second "soft-failure" case is much worse, as it could lead to unexpected behavior in the deployed chart application, likely requiring a much more involved debugging requirement and confusing user experience.

Author: The premise behind the `minimumHelmVersion` field, is that the chart requires a feature present only in newer versions of Helm. If a user wants that chart, they need that newer version of Helm :) 

If they also want an old version of Kubernetes, then yeah, they are perhaps out of luck. But two things can be said here:
- `minimumHelmVersion` isn't changing the fact that a newer version of Helm is needed. It is just making the error upfront (as this section points out)
-  The version skew is the "guaranteed" compatibility matrix. Versions tend to work outside of this range. So they user may be okay with the new Helm / old Kubernetes (it just isn't "guaranteed").

So I don't think it will "force" users to upgrade Kubernetes versions. The only situation where an upgrade would be required, is if a user does want a new chart version, which requires a new Helm version. And the chart or Helm is incompatible with the old Kubernetes version.

Supporting a simple SemVer based `minimumHelmVersion` field in chart's `Chart.yaml` is thought to be a simple way to provide forward compatibility guarantees.

## Rationale

It is thought that over time, as Helm extends its functionality, there will be greater scope for forward compatibility errors to present themselves.

And in particular, the mechanism that describes the minimum feature set must be built into and understood by prior versions of Helm.

Utilizing a simple minimum version SemVer field in a chart is thought to be a simple and succinct way to enable forward compatibility guarantees for future versions of Helm.

Similar to the way the Go toolchains allow forward compatibility guarentees via [`go.mod`][1].

## Specification

`Chart.yaml` will include an optional field `minimumHelmVersion`, which is a [SemVer][2] string for the version of Helm required for the chart to operate:

```
minimumHelmVersion: <MAJOR>[.<MINOR>[.PATCH]]
```

The minor and patch fields are optional.
And will be inferred as zeros if not specified.

> When loading a chart, iff the `minimumHelmVersion` exists, Helm will verify its version exceeds or equals the version constraint specified in the field.

Author: This is one of my favorite contractions :) iff -> if and only

https://www.merriam-webster.com/dictionary/iff

ie. the above reads like:

> When loading a chart, if and only if the `minimumHelmVersion` exists, Helm will verify its version exceeds or equals the version constraint specified in the field.

> When loading a chart, iff the `minimumHelmVersion` exists, Helm will verify its version exceeds or equals the version constraint specified in the field.

Author: ok, fair. will drop usage of `iff`

When "strict" loading of `Chart.yaml` is used: ie. the loading of `Chart.yaml` errors when unknown fields are present, Helm will first 'peek' into `Chart.yaml` and extract the `minimumHelmVersion` field (iff it exists) and perform the version constraint check at this point.
This is to provide better UX that simply erroring on any potential new fields in `Chart.yaml` that are unknown to the current version of Helm.

> `helm create` will pre-fill Chart.yaml's `minimumHelmVersion` field with the current version of Helm.

Author: I've remove this. Goal was to keep new charts requiing new versions of Helm. But agree, this is too agressive.

The default starter chart should contain the `minimumHelmVersion` version field, if the chart requires that minimum version of Helm to operate.

While this may not be strictly necessary, it will provide a backstop for features that may be included in the chart by the user for their current Helm version or `helm create`.
The user can remove/reduce the specified `minimumHelmVersion` if desired.

> `helm lint` will produce a warning, if the current version of Helm is newer than the minimum version constraint.

Author: I agree this is too aggressive. 

I did originally mean _newer_, with the intention as follows. Will address in https://github.com/helm/community/pull/370#discussion_r1852354989

Or the `minimumHelmVersion` field is unset for the given chart.

> This is to encourage users to update the minimum version to currently utilized version of Helm (that the user is using to develop the chart).

Author: I've removed `helm lint` reporting missing or old `minimumChartVersion` field. While users keeping Helm up to date would be good, this probably isn't the right way to "encourage" them to do so.

## Backwards compatibility

As `minimumHelmVersion` is an optional field, it will be inferred to be unset in existing charts.

> And when unset, Helm will retain existing behavior (won't implement any checks)

Reviewer: ```suggestion
And when unset, Helm will retain existing behavior (no check will be implemented)
```

> Older versions of Helm which don't understand the `minimumHelmVersion` field will actually ignore the field even if set (as they load `Chart.yaml` ignoring unknown fields).

Reviewer: ```suggestion
Older versions of Helm which do not understand the `minimumHelmVersion` field will ignore the field even if set (as they load `Chart.yaml` ignoring unknown fields).
```

However, there isn't anything that can practically be done to address this encoded behavior.

## Security implications

None

## How to teach this

Update docs for `Chart.yaml`

## Reference implementation

N/A

## Rejected ideas

**Generic version contraints:**
It is thought a "minimum" constraint is preferred over a generic "constraint" system.
ie. where a user could list multiple version constraints to be satisfied
Similar to e.g. [pip's requirement-specifiers][3].

> This is because Helm makes great effort to remain [backwards compatible][4].

Reviewer: ```suggestion
This is because Helm makes great effort to remain [backwards compatibility][4].
```

And so it is presumed unnecessary that a user would need to specify anything but a single minimum version constraint.
And in the future when Helm does make breaking changes to charts, it is expected the chart API version will be incremented to explicitly signify this new incompatiblity.

**Capabilities:**
Providing a capabilities based system.
Rather than just directly inferring Helm capabilities from its version.

A `Chart.yaml` could describe its required features ie. capabilities it expects/requires the Helm version to provide.

> From which Helm could inspect and matches against features/capabilities the in use version does provide.

Reviewer: ```suggestion
Helm could inspect and match against features/capabilities that the specific version does provide.
```

However, since core Helm capabilities are only every additive due to Helm's [backwards-compatibility rules][4], a capabilities based system is redundant.

*NB: at the time of writing, plugins for extending Helm's functionality are being heavily discussed.

> And while plugins will likely allow extending capabilities independent of the Helm version.

Reviewer: Combine with the following sentence? Its currently a fragment

Author: oops, right
